src/options/template.rs

```diff
@@ -29,83 +31,58 @@ impl TemplateOpt {
                 .interact_on_opt(&Term::stderr())?;
 
             match selection {
-                Some(index) => Ok(languages[index]),
+                Some(index) => Ok(languages[index].clone()),
                 None => Err(RoverError::new(anyhow!("No language selected"))),
             }
         }
     }
 }
 
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
-pub struct GithubTemplate {
-    pub id: &'static str,
-    pub git_url: &'static str,
-    pub display: &'static str,
-    pub language: ProjectLanguage,
-}
-
-impl GithubTemplate {
-    pub(crate) fn repo_slug(&self) -> Result<&'static str> {
-        self.git_url
-            .split('/')
-            .last()
-            .ok_or_else(|| anyhow!("Could not determine tarball path.").into())
-    }
-
-    pub(crate) fn extract_github_tarball(
-        &self,
-        template_path: &Utf8PathBuf,
-        client: &reqwest::blocking::Client,
-    ) -> Result<()> {
-        let download_dir = tempdir::TempDir::new(self.id)?;
-        let download_dir_path = Utf8PathBuf::try_from(download_dir.into_path())?;
-        let git_repo_slug = self.repo_slug()?;
-        let tarball_path = download_dir_path.join(format!("{}.tar.gz", git_repo_slug));
-        let tarball_url = format!("{}/archive/refs/heads/main.tar.gz", &self.git_url);
-        let mut f = std::fs::File::create(&tarball_path)?;
-        eprintln!("Downloading {}", &self.git_url);
-        eprintln!("\tfrom {}", tarball_url);
-        let response_bytes = client
-            .get(tarball_url)
-            .header(reqwest::header::USER_AGENT, "rover-client")
-            .header(reqwest::header::ACCEPT, "application/octet-stream")
-            .send()?
-            .error_for_status()?
-            .bytes()?;
-        f.write_all(&response_bytes[..])?;
-        f.sync_all()?;
-        let f = std::fs::File::open(&tarball_path)?;
-        let tar = flate2::read::GzDecoder::new(f);
-        let mut archive = tar::Archive::new(tar);
-        archive
-            .unpack(&template_path)
-            .with_context(|| format!("could not unpack tarball to '{}'", &template_path))?;
-
-        let tar_path = template_path.join(format!("{}-main", git_repo_slug));
-
-        // The unpacked tar will be in the folder{git_repo_id}-{branch}
+pub(crate) fn extract_github_tarball(
+    download_url: Url,
+    template_path: &Utf8PathBuf,
+    client: &reqwest::blocking::Client,
+) -> Result<()> {
+    let download_dir = tempdir::TempDir::new("rover-template")?;
+    let download_dir_path = Utf8PathBuf::try_from(download_dir.into_path())?;
+    let file_name = download_url
+        .as_str()
+        .split('/')
+        .last()
+        .ok_or_else(|| anyhow!("Could not determine tarball path."))?;
```

Comment:
I think we should have the API itself hold the `.tar.gz` endpoints directly, that way we can just use the data without parsing out the string (and potentially support non-github git providers down the line, or even just S3 bucket endpoints that may or may not follow the same URL conventions regarding tarball downloads)

Reply:
The `.tar.gz` endpoint is coming from the API (`download_url`) but the actual file name of the file being downloaded has to be determined—so I do this by getting the final component of the URI. We could probably just set our own filename somehow instead...

Reply:
oh i see. i guess for that we could maybe just use the name of the template directory they're trying to create but with a `.tar.gz` suffix? so if i ran `rover template use hello-world` and selected the apollo server typescript template, it would download the tarball to `hello-world.tar.gz` and extract it to `hello-world`?

Reply:
There was some weird behavior with a nested folder inside of the extracted archive being named something deterministic based on the download URI. Not sure if that's a GitHub thing or a tar thing. Maybe we can pair to sort that out together?

Reply:
Sorted in pairing!